src/3d/qgsfgutils.h

```diff
@@ -0,0 +1,81 @@
+/***************************************************************************
+  qgsfgutils.h
+  --------------------------------------
+  Date                 : August 2024
+  Copyright            : (C) 2024 by Mike Krus / Benoit De Mezzo
+  Email                : mike dot krus at kdab dot com / benoit dot de dot mezzo at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSFG_UTILS_H
+#define QGSFG_UTILS_H
+
+#include <QWindow>
+#include <Qt3DRender/QCamera>
+#include <Qt3DRender/QRenderSurfaceSelector>
+#include <Qt3DRender/QViewport>
+#include <Qt3DRender/QCameraSelector>
+#include <Qt3DRender/QLayerFilter>
+#include <Qt3DRender/QLayer>
+#include <Qt3DRender/QRenderTargetSelector>
+#include <Qt3DRender/QRenderTarget>
+#include <Qt3DRender/QTexture>
+#include <Qt3DRender/QClearBuffers>
+#include <Qt3DRender/QParameter>
+#include <Qt3DRender/QFrustumCulling>
+#include <Qt3DRender/QRenderStateSet>
+#include <Qt3DRender/QDepthTest>
+#include <Qt3DRender/QCullFace>
+#include <Qt3DRender/QPolygonOffset>
+#include <Qt3DRender/QRenderCapture>
+#include <Qt3DRender/QDebugOverlay>
+
+
+#define SIP_NO_FILE
+
+/**
+ * \ingroup 3d
+ * \brief Util class to dump Qt3D framegraph or scenegraph.
+ * \since QGIS 3.40
+ */
+class QgsFgUtils
+{
+  public:
+
+    //! Returns a tree view of the scene graph starting from \a node. The object ids will be given relatively to \a lowestId.
+    QStringList dumpSceneGraph( const Qt3DCore::QNode *node, int lowestId = 0 );
+
+    //! Returns a tree view of the frame graph starting from \a node. The object ids will be given relatively to \a lowestId.
+    QStringList dumpFrameGraph( const Qt3DCore::QNode *node, int lowestId = 0 );
+
+  private:
+    QString formatIdName( quint64 id, const QString &name );
+
+    QString formatIdName( const Qt3DRender::QAbstractTexture *texture );
+
+    QString formatNode( const Qt3DCore::QNode *n );
+
+    QString formatList( const QStringList &lst );
+
+    QString formatLongList( const QStringList &lst, int level );
+
+    QString formatField( const QString &name, const QString &value );
+
+    QString dumpSGEntity( const Qt3DCore::QEntity *n, int level );
+
+    QStringList dumpSG( const Qt3DCore::QNode *n, int level = 0 );
+
+    QString dumpFGNode( const Qt3DRender::QFrameGraphNode *n );
+
+    QStringList dumpFG( const Qt3DCore::QNode *n, int level = 0 );
+
+    int mLowestId = 0;
```

Comment:
I think it'd be cleaner to make dumpSceneGraph and dumpFrameGraph static (matching the behaviour of other Qgs...Utils classes). Instead of a member variable for lowest ID we could just pass it to the private methods directly (or alternatively, make a private struct:

```
   struct Context 
   {
       int lowestId = 0;
    }
```

and pass that to the private methods, which would be more future-proof).

Reply:
Indeed! But I was too lazy to change all the private functions :)